hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java

```diff
@@ -451,188 +378,156 @@ private String getInventoryId(NamedObject no) {
         // We first must persist metric types then resource types (remember, there are no hierarchies,
         // all metric types and resource types are peers to one another).
-        List<ResourceType<L>> allResourceTypes = resourceTypeManager.getResourceTypesBreadthFirst()
-                .stream()
-                .filter(rt -> rt.isPersisted() == false)
-                .collect(Collectors.toList());
-        Map<MeasurementType<L>, Offline<org.hawkular.inventory.api.model.MetricType.Blueprint>> mtBlueprints;
-        mtBlueprints = bldr.buildMetrics(allResourceTypes);
-        mtBlueprints.forEach((mt, bp) -> performMetricTypeSync(bp, tenantIdToUse));
-
-        Map<ResourceType<L>, Offline<org.hawkular.inventory.api.model.ResourceType.Blueprint>> rtBlueprints;
-        rtBlueprints = bldr.build(allResourceTypes);
-        rtBlueprints.forEach((rt, bp) -> performResourceTypeSync(bp, tenantIdToUse));
-
-        // indicate we persisted the resource types
-        allResourceTypes.forEach(rt -> rt.setPersisted(true));
+        event.getResourceTypeManager().ifPresent(resourceTypeManager -> {
+            List<ResourceType<L>> allResourceTypes = resourceTypeManager.getResourceTypesBreadthFirst();
+            getMetricTypesToSync(allResourceTypes).forEach(mt -> {
+                log.infof("Updating metric type: %s", mt.getID().getIDString());
+                performMetricTypeSync(bldr.buildMetricType(mt), tenantIdToUse);
+                // indicate we persisted the resource
+                mt.setPersistedTime(timestamp);
+            });
+
+            allResourceTypes.stream()
+                    .filter(this::needToRefresh)
+                    .forEach(rt -> {
+                        log.infof("Updating resource type: %s", rt.getID().getIDString());
+                        performResourceTypeSync(bldr.buildResourceType(rt), tenantIdToUse);
+                        // indicate we persisted the resource
+                        rt.setPersistedTime(timestamp);
+                    });
+        });
 
         // build the JSON blueprints for the sync resource requests
         // Note that it is possible for a endpoint to define multiple root resources.
         // We have to sync each root resource separately.
-        Map<Resource<L>, Offline<org.hawkular.inventory.api.model.Resource.Blueprint>> rBlueprints;
-        rBlueprints = bldr.build(resourceManager);
-        rBlueprints.forEach((r, bp) -> performResourceSync(bp, tenantIdToUse, resourceManager.size(r)));
+        Set<ID> addedOrModifiedIds = event.getAddedOrModifiedRootResources().stream()
+                .map(IDObject::getID)
+                .collect(Collectors.toSet());
+        resourceManager.getRootResources().forEach(r -> {
+            // Ignore unmodified resources that doesn't need to be refreshed
+            if (addedOrModifiedIds.contains(r.getID()) || needToRefresh(r)) {
+                log.infof("Updating root resource: %s", r.getID().getIDString());
```

Comment:
@jmazzitelli are you fine with those 3 log.info (also line 444 and 435) that happen every time inventory storage is updated? That can be interesting for tracing http calls, but if you prefer I can put to debug level (or remove).